File: BT.C++/Buổi 2 C++/2.2.4.cpp
#include <iostream>
#include <vector>
using namespace std;

// Hàm Bubble Sort
void bubbleSort(vector<int>& arr) {
    int n = arr.size();
    bool swapped;

    for (int i = 0; i < n - 1; i++) { // Sửa lỗi cú pháp (n - 1 thay vì n 1)
        swapped = false;

        for (int j = 0; j < n - 1 - i; j++) { // Sửa lỗi cú pháp (n-1-i thay vì n-1- 1)
            if (arr[j] > arr[j + 1]) {
                // Hoán đổi
                int temp = arr[j]; // Sửa lỗi (temp arr[j] thành temp = arr[j])
                arr[j] = arr[j + 1];
                arr[j + 1] = temp;
                swapped = true;
            }
        }

        // Nếu không có hoán đổi nào, mảng đã được sắp xếp
        if (!swapped)
            break;
    }
}

int main() {
    vector<int> arr = {5, 2, 9, 1, 5}; // Sửa lỗi cú pháp ({5, 2, 9, 1, 5} thay vì (5, 2, 9, 1, 5))

    bubbleSort(arr);

    cout << "Ket qua sau Bubble Sort: ";
    for (int x : arr) // Sửa lỗi cú pháp
        cout << x << " ";
    cout << endl;

    return 0;
}

File: BT.Python/Buổi 2 Python/2.2.4.py
def bubble_sort(arr):
    n = len(arr)

    for i in range(n - 1):
        swapped = False
        for j in range(n - 1 - i):  # Sửa lỗi (n-11 -> n-1-i)
            if arr[j] > arr[j + 1]:
                # Hoán đổi
                arr[j], arr[j + 1] = arr[j + 1], arr[j]
                swapped = True

        # Nếu không có hoán đổi nào, mảng đã được sắp xếp
        if not swapped:
            break


if __name__ == "__main__":
    arr = [5, 2, 9, 1, 5]  # Sửa cú pháp khai báo mảng
    bubble_sort(arr)
    print("Kết quả sau Bubble Sort:", arr)  # Sửa cú pháp in kết quả
